tests/src/unit-custom-base-class.cpp

```diff
@@ -0,0 +1,357 @@
+/*
+    __ _____ _____ _____
+ __|  |   __|     |   | |  JSON for Modern C++ (test suite)
+|  |  |__   |  |  | | | |  version 3.10.2
+|_____|_____|_____|_|___|  https://github.com/nlohmann/json
+
+Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+SPDX-License-Identifier: MIT
+Copyright (c) 2013-2019 Niels Lohmann <http://nlohmann.me>.
+
+Permission is hereby  granted, free of charge, to any  person obtaining a copy
+of this software and associated  documentation files (the "Software"), to deal
+in the Software  without restriction, including without  limitation the rights
+to  use, copy,  modify, merge,  publish, distribute,  sublicense, and/or  sell
+copies  of  the Software,  and  to  permit persons  to  whom  the Software  is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE  IS PROVIDED "AS  IS", WITHOUT WARRANTY  OF ANY KIND,  EXPRESS OR
+IMPLIED,  INCLUDING BUT  NOT  LIMITED TO  THE  WARRANTIES OF  MERCHANTABILITY,
+FITNESS FOR  A PARTICULAR PURPOSE AND  NONINFRINGEMENT. IN NO EVENT  SHALL THE
+AUTHORS  OR COPYRIGHT  HOLDERS  BE  LIABLE FOR  ANY  CLAIM,  DAMAGES OR  OTHER
+LIABILITY, WHETHER IN AN ACTION OF  CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "doctest_compatibility.h"
+
+#include <nlohmann/json.hpp>
+
+// Test extending nlohmann::json by using a custom base class.
+// Add some metadata to each node and test the behaviour of copy / move
+template<class MetaDataType>
+class json_metadata
+{
+  public:
+    using metadata_t = MetaDataType;
+    metadata_t& metadata()
+    {
+        return m_metadata;
+    }
+    const metadata_t& metadata() const
+    {
+        return m_metadata;
+    }
+  private:
+    metadata_t m_metadata = {};
+};
+
+template<class T>
+using json_with_metadata =
+    nlohmann::basic_json <
+    std::map,
+    std::vector,
+    std::string,
+    bool,
+    std::int64_t,
+    std::uint64_t,
+    double,
+    std::allocator,
+    nlohmann::adl_serializer,
+    std::vector<std::uint8_t>,
+    json_metadata<T>
+    >;
```

Comment:
Yikes. Should we consider adding a template alias to simplify the default use cases?

Something like this:
```cpp
template<typename CustomBaseClass, ...>
using json_with_base = basic_json<..., CustomBaseClass>;
```

(Maybe in a separate PR.)

Reply:
I agree, the current syntax is ugly.
If we want to add an alias, it would always have to be consistent with the template for `basic_json`

e.g.:

    template<class CustomBaseClass,
            template<typename U, typename V, typename... Args> class ObjectType =
            std::map,
            template<typename U, typename... Args> class ArrayType = std::vector,
            class StringType = std::string, class BooleanType = bool,
            class NumberIntegerType = std::int64_t,
            class NumberUnsignedType = std::uint64_t,
            class NumberFloatType = double,
            template<typename U> class AllocatorType = std::allocator,
            template<typename T, typename SFINAE = void> class JSONSerializer =
            adl_serializer,
            class BinaryType = std::vector<std::uint8_t>>
    using json_with_base = basic_json<ObjectType, ArrayType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType, CustomBaseClass>;

So we always would have to duplicate and update everything from the base template.

There is a way to circumvent this. We could add `basic_json::with_base_t<T>`. This alias would use the parameters from the current type. In this case the namespace level alias could be:

    template<class CustomBaseClass, class...Ts>
    using json_with_base = basic_json<Ts...>::with_base_t<CustomBaseClass>;

Now we would only have to update `basic_json::with_base_t` and not the defaults.

Since this change is not trivial and might need some discussion, i would leave it to a different PR.

Reply:
@barcode You can mark this as resolved.